docs/upgrading_to_nameko_3.rst

```diff
@@ -0,0 +1,347 @@
+
+    container = ServiceContainer(Service)
+    container.start()
+    # AMQP extensions will connect as someuser to RabbitMQ broker at somehost ...
+    container.stop()
+
+Same for the runner:
+
+.. code-block:: python
+
+    from nameko.runners import ServiceRunner
+
+    nameko.config["AMQP_URI"] = "pyamqp://someuser:*****@somehost/"
+
+    runner = ServiceRunner()
+    runner.add_service(ServiceA)
+    runner.add_service(ServiceB)
+
+    runner.start()
+    # AMQP extensions will connect as someuser to RabbitMQ broker at somehost ...
+    runner.stop()
+
+Other configuration related changes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+* The `--broker`` CLI option is deprecated in favour of `--define` and `--config`. TODO: doesn’t actually raise a warning anymore
+* The built-in Config dependency provider is deprecated as the config can be accessed and read directly.
+
+Testing services
+----------------
+
+In Nameko 2, configuration was always explicitly passed as an object, which made it easy to manage in tests. The global config object needs some different ways of managing configuration in tests.
+
+Config patching
+~~~~~~~~~~~~~~~
+
+The global `config` object has a `patch` helper object which can be used as a context manager or as a decorator. Here is an example of patching the config object in a pytest fixture:
+
+.. code-block:: python
+
+    import nameko
+    import pytest
+
+    @pytest.fixture
+    def memory_rabbit_config():
+        with nameko.config.patch({"AMQP_URI": "memory://"}):
+            yield
+
+While this fixture is in place, any code accessing `config['AMQP_URI']` will receive the value `"memory://"`.
+
+Changes to the `container_factory` and `runner_factory` fixtures.
+
+In Nameko 2, the `container_factory` and `runner_factory` fixtures accepted `config` as a required argument. In Nameko 3, it is optional. If passed, `nameko.config.patch` is used to wrap running container. The patch is applied before container start and returned it back to its original value when exiting the test:
+
+.. code-block:: python
+
+    def test_with_container_factory(container_factory):
+        container = container_factory(Service, config={"AMQP_URI": "..."})
+        container.start()  # <- starts config patching scope, ending it when container
+                        #    stops (exit of the container_factory fixture)
+        # ...
+
+    def test_with_runner_factory(runner_factory):
+        runner = runner_factory(config={"AMQP_URI": "..."}, ServiceX, ServiceY)
+        runner.start()  # <- starts config patching scope, ending it when runner
+                        #    stops (exit of the runner_factory fixture)
+        # ...
+
+The optional argument provides backward compatibility so that tests for Nameko extensions can be compatible with both major versions.
+
+When creating or upgrading existing services to run on Nameko 3, it is recommended to use the new config patch pattern instead:
```

Comment:
Because using the alternative to `config.patch()` is deprecated, I would reformulate this in a more directive manner:

> When creating or upgrading services to run on Nameko 3, manipulate config options through the new `config.patch` pattern like so: